libbeat/common/kubernetes/util.go

```diff
@@ -1,68 +1,123 @@
+func DiscoverKubernetesNode(host string, inCluster bool, clientset *kubernetes.Clientset) (node string) {
+	if host != "" {
+		logp.Info("kubernetes: Using node %s provided in the config", host)
+		return host
+	}
+
+	if inCluster {
+		ns := inClusterNamespace()
+		podName, err := os.Hostname()
+		if err != nil {
+			logp.Err("kubernetes: Couldn't get hostname as beat pod name in cluster with error: ", err.Error())
+			return defaultNode
 		}
-		client, err = k8s.NewClient(&config)
+		logp.Info("kubernetes: Using pod name %s and namespace %s to discover kubernetes node", podName, ns)
+		pod, err := clientset.CoreV1().Pods(inClusterNamespace()).Get(podName, v1.GetOptions{})
 		if err != nil {
-			return nil, err
+			logp.Err("kubernetes: Querying for pod failed with error: ", err.Error())
+			return defaultNode
 		}
+		logp.Info("kubernetes: Using node %s discovered by in cluster pod node query", pod.Spec.NodeSelector)
+		return pod.Spec.NodeName
 	}
 
-	return client, nil
+	mid := machineID()
+	if mid == "" {
+		logp.Err("kubernetes: Couldn't collect info from any of the files in /etc/machine-id /var/lib/dbus/machine-id")
+		return defaultNode
+	}
+
+	nodes, err := clientset.CoreV1().Nodes().List(v1.ListOptions{})
+	if err != nil {
+		logp.Err("kubernetes: Querying for nodes failed with error: ", err.Error())
+		return defaultNode
+	}
+	for _, n := range nodes.Items {
+		if n.Status.NodeInfo.MachineID == mid {
+			logp.Info("kubernetes: Using node %s discovered by machine-id matching", n.GetName())
+			return n.GetName()
+		}
+	}
+
+	logp.Warn("kubernetes: Couldn't discover node, using localhost as default")
+	return defaultNode
 }
 
-// DiscoverKubernetesNode figures out the Kubernetes host to use. If host is provided in the config
-// use it directly. Else use hostname of the pod which is the Pod ID to query the Pod and get the Node
-// name from the specification. Else, return localhost as a default.
-func DiscoverKubernetesNode(host string, client *k8s.Client) string {
-	ctx := context.Background()
-	if host == "" {
-		podName := os.Getenv("HOSTNAME")
-		logp.Info("Using pod name %s and namespace %s", podName, client.Namespace)
-		if podName == "localhost" {
-			host = "localhost"
-		} else {
-			pod, err := client.CoreV1().GetPod(ctx, podName, client.Namespace)
-			if err != nil {
-				logp.Err("Querying for pod failed with error: ", err.Error())
-				logp.Info("Unable to find pod, setting host to localhost")
-				host = "localhost"
-			} else {
-				host = pod.Spec.GetNodeName()
-			}
+// inClusterNamespace gets namespace from serviceaccount when beat is in cluster.
+// code borrowed from client-go with some changes.
+func inClusterNamespace() string {
+	// This way assumes you've set the POD_NAMESPACE environment variable using the downward API.
+	// This check has to be done first for backwards compatibility with the way InClusterConfig was originally set up
+	if ns := os.Getenv("POD_NAMESPACE"); ns != "" {
```

Comment:
We stopped using this, in favor of `client.Namespace`, what's the motivation for this change?

Reply:
Since `k8s.io/client-go` doesn't have a field `Namespace` in it, so `client.Namespace` is kind like a tricky in `ericchiang/k8s` I think. This func is borrowed from client-go, and after rethink about it, I think user should not set namespace according downward API since if beat is in cluster, then it should be auto detected, if beat is outside cluster, this checking is meaningless. Will remove it from new PR.